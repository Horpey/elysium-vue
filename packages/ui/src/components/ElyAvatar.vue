<script setup lang="ts">
import type { ElyAvatarSizeKey } from '~/types/ElyAvatar'

interface Props {
  src?: string
  alt: string
  size?: ElyAvatarSizeKey
  chip?: boolean
}

const props = withDefaults(defineProps<Props>(), {
  src: '',
  alt: '',
  size: 'MEDIUM',
})

const classes = computed(() => {
  let sizeClass = ''

  switch (props.size) {
    case 'SMALL':
      sizeClass = 'h-6 w-6'
      break
    case 'MEDIUM':
      sizeClass = 'h-10 w-10'
      break
    case 'LARGE':
      sizeClass = 'h-16 w-16'
      break
  }

  return sizeClass
})
</script>

<template>
  <span
    class="relative m-2 inline-flex items-center justify-center border-koromiko-700 bg-koromiko-800/10 text-koromiko-700 ring-1 ring-koromiko-700 ring-offset-2"
    :class="classes"
  >
    <img v-if="src" :src="src" :alt="alt">

    <span v-else class="font-medium uppercase leading-none">{{
      alt
        .split(' ')
        .map((word) => word[0])
        .join('')
    }}</span>

    <span
      v-if="chip" class="absolute right-0 top-0 block -translate-y-1/2 translate-x-1/2 bg-green-400 ring-2 ring-white"
      :class="{
        'size-1.5': size === 'SMALL',
        'size-2': size === 'MEDIUM',
        'size-3': size === 'LARGE',
      }"
    />
  </span>
</template>
