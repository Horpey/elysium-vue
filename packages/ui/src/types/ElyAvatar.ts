export const ELY_AVATAR_SIZE = {
  LARGE: 'large',
  MEDIUM: 'medium',
  SMALL: 'small',
} as const

export type ElyAvatarSizeKey = keyof typeof ELY_AVATAR_SIZE
