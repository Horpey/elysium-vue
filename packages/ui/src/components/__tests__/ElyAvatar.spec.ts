import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import ElyAvatar from '../ElyAvatar.vue'

describe('elyAvatar', () => {
  it('renders with default props', async () => {
    const wrapper = mount(ElyAvatar, {
      props: {
        src: 'https://example.com/avatar.jpg',
        alt: 'John Doe',
      },
    })

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)

    // Assert default values
    expect(wrapper.find('.relative.m-2.inline-flex.items-center.justify-center.border-koromiko-700').exists()).toBe(true)
    expect(wrapper.find('img').exists()).toBe(true)
    expect(wrapper.find('img').attributes('src')).toBe('https://example.com/avatar.jpg')
    expect(wrapper.find('img').attributes('alt')).toBe('John Doe')
    expect(wrapper.find('.h-10.w-10').exists()).toBe(true) // Medium size class

    // // Assert that chip is not rendered
    expect(wrapper.find('.absolute.right-0.top-0.block').exists()).toBe(false)

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders without src and with chip', async () => {
    const wrapper = mount(ElyAvatar, {
      props: {
        alt: 'John Doe',
        chip: true,
      },
    })

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)
    expect(wrapper.find('img').exists()).toBe(false)
    expect(wrapper.find('.font-medium.uppercase.leading-none').text()).toBe('JD') // Initials from alt
    expect(wrapper.find('.absolute.right-0.top-0.block').exists()).toBe(true)
    expect(wrapper.find('.h-2.w-2').exists()).toBe(true) // Medium size class

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })
})
