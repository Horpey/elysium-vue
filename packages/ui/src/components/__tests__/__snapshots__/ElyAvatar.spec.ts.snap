// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`elyAvatar > renders with default props 1`] = `"<span class="relative m-2 inline-flex items-center justify-center border-koromiko-700 bg-koromiko-800/10 text-koromiko-700 ring-1 ring-koromiko-700 ring-offset-2 h-10 w-10"><img src="https://example.com/avatar.jpg" alt="John Doe"><!--v-if--></span>"`;

exports[`elyAvatar > renders without src and with chip 1`] = `"<span class="relative m-2 inline-flex items-center justify-center border-koromiko-700 bg-koromiko-800/10 text-koromiko-700 ring-1 ring-koromiko-700 ring-offset-2 h-10 w-10"><span class="font-medium uppercase leading-none">JD</span><span class="absolute right-0 top-0 block -translate-y-1/2 translate-x-1/2 bg-green-400 ring-2 ring-white size-2"></span></span>"`;
